### DISCUSSION TITLE
Skeleton Loading States

### DISCUSSION THREAD

**Discussion by vankhoawin** [twilio-labs/paste#448] [Ideas] [answered]
The Spinner component is useful for showing that sections of a page are loading, but there are some reasons why skeleton loading states could possibly provide better UX than Spinners (I won't list them, there's enough articles on Google for that). But the main point I would say is to "bring placeholders closer to their content."

Examples:

**Medium**
![image](https://user-images.githubusercontent.com/6936736/82083237-3d8ef680-969e-11ea-912d-b14a58802e4d.png)

**Google Drive**
![image](https://user-images.githubusercontent.com/6936736/82083250-4384d780-969e-11ea-8f87-a3dee11ebc4d.png)

**Comment by aayushpi** [comment]
Thank you @vnguyen94! Personally, I'm in favor of this feature. Let's keep this discussion open and solicit demand for it so we can consider how we may prioritize it. It may also be a good contribution. I believe @GiantRobots has some examples he'd like to add too

**Comment by GiantRobots** [comment]
<img width="1413" alt="Screen Shot 2020-05-15 at 1 59 35 PM" src="https://user-images.githubusercontent.com/1903839/82095654-573b3880-96b4-11ea-88d8-2761a408e983.png">
<img width="1413" alt="Screen Shot 2020-05-15 at 1 58 56 PM" src="https://user-images.githubusercontent.com/1903839/82095671-5e624680-96b4-11ea-91a6-0b16fc6f9bd8.png">
<img width="1413" alt="Screen Shot 2020-05-15 at 1 58 22 PM" src="https://user-images.githubusercontent.com/1903839/82095690-68844500-96b4-11ea-8001-9fd0678acf16.png">

**Comment by GiantRobots** [comment]
Those mostly show some skeleton ui's with cards/tables. We have 3 loaders for SG the fragment loader, the SG loader, and the table loader each with differing rules on how to implement them located in the styleguide 

http://styleguide.sendgrid.com/loaders.html
http://styleguide.sendgrid.com/tables.html#tables-loading

I think it helps a lot in the feeling that you navigated instead of throwing a large "page" loader instead of the skeleton pages. It can also help with reflow issues.

Some of the issues we've experienced around them however include scopes/flags/user perms on a  feature that show the skeleton  outline and then we use a card to indicate they don't have access to the feature. That experience can be jarring at times but we normally have to deal with it because of race conditions for when http requests normally resolve.

**Comment by GiantRobots** [comment]
There's some interesting conversation about how applications load their data in the react docs

https://reactjs.org/docs/concurrent-mode-suspense.html#traditional-approaches-vs-suspense

I think we  use the `fetch-on-render` paradigm which leads to lots of waterfall http requests as well as waterfall layouts.

More fun thoughts from the suspense doc.


> Note how we eliminated the if (...) “is loading” checks from our components. This doesn’t only remove boilerplate code, but it also simplifies making quick design changes. For example, if we wanted profile details and posts to always “pop in” together, we could delete the <Suspense> boundary between them. Or we could make them independent from each other by giving each its own <Suspense> boundary. Suspense lets us change the granularity of our loading states and orchestrate their sequencing without invasive changes to our code.


Using the suspense api will eventually allow  us to pull the loading parts of our components out and decide where the barriers are. We may want to provide insight on how to handle the various application loads users of the  library will encounter to keep our UI eventually consistent

**Comment by SiTaggart** [comment, marked as answer]
@GiantRobots @vnguyen94 Referencing prior art, would something like Chakra UIs skeleton component fit the use cases you are thinking?

https://chakra-ui.com/skeleton

**Reply by vankhoawin** [reply]
For our team's use cases, I think so yes; our team doesn't have any out-of-the-ordinary flows (Studio/Flex come to mind). We'd use skeletons as placeholders for loading lists and sections, so having a base Skeleton component would let us compose those larger molecule components.

**Reply by GiantRobots** [reply]
I like the API for that but we should keep an eye on the suspense API and how to build skeletons for that.

If we use the same API from Chakra for our own Skeletons

```javascript
import { Box } from "@chakra-ui/core";

function Card() {
  const { data, loading, error } = useRemoteData();
  if (error) return <Box children="error" />;
  return (
    <Box>
      <Skeleton isLoaded={!loading}>
        <Heading>{data.title}</Heading>
      </Skeleton>
    </Box>
  );
}
```

And use it for Skeletons/Loaders with suspense


```javascript

function Card() {
  return (
    <Box>
      <Suspense fallback={<Skeleton>Fake Child Components so it renders correctly</Skeleton>}>
        <ComponentThatDoesFetching>My Awesome component</ComponentThatDoesFetching>
      </Suspense>
    </Box>
  );
}
```

We get some weirdness with layout. It might be better to have Card/Box Skelly, Table Skelly, etc. for the various use cases.


There was some weirdness with the implementation for SendGrid tables where it required the tables to be empty to get the loading to trigger because it was implemented in css 100%. So sometimes we'd get weird bugs where prettier added a space and a row component would stop loading (I would prefer an explicit API for those states) or specific components for them.

So I guess my recommendation is some combination of Skeleton with isLoaded for current ease of use, and maybe in the future a suspense wrapper.



After thought, we can also implement the Skeleton loader like this


```javascript
const SuspenseSkelly = ({children}) => {
<Suspense fallback={<Skeleton>{default size}</Skeleton>}>
       {children}
</Suspense>
}

// Using it would look like

<SuspenseSkelly>
      <MyComponentWithSuspense />
</SuspenseSkelly>

```
